src/coprocessor/dag/executor/scanner.rs

```diff
@@ -120,47 +113,73 @@ impl<S: Snapshot> Scanner<S> {
             );
         }
 
-        self.seek_key = match (self.scan_mode, self.scan_on) {
-            (ScanMode::Forward, _) => {
-                let mut seek_key = key.clone();
-                util::convert_to_prefix_next(&mut seek_key);
-                seek_key
-            }
-            (ScanMode::Backward, ScanOn::Table) => box_try!(truncate_as_row_key(&key)).to_vec(),
-            (ScanMode::Backward, ScanOn::Index) => key.clone(),
-            _ => unreachable!(),
-        };
+        // `Vec::clear()` produce 2 more instructions than `set_len(0)`, so we directly use
+        // `set_len()` here.
+        unsafe {
+            self.last_scanned_key.set_len(0);
+        }
+        self.last_scanned_key.extend_from_slice(key.as_slice());
 
         Ok(Some((key, value)))
     }
 
     pub fn start_scan(&self, range: &mut KeyRange) {
         assert!(!self.no_more);
-        let cur_seek_key = self.seek_key.clone();
-        match self.scan_mode {
-            ScanMode::Forward => range.set_start(cur_seek_key),
-            ScanMode::Backward => range.set_end(cur_seek_key),
-            _ => unreachable!(),
-        };
+        if self.last_scanned_key.is_empty() {
+            // Happens when new() -> start_scan().
+            match self.scan_mode {
+                ScanMode::Forward => range.set_start(self.range.get_start().to_owned()),
+                ScanMode::Backward => range.set_end(self.range.get_end().to_owned()),
+                _ => unreachable!(),
+            }
+        } else {
+            // Happens when new() -> start_scan() -> next_row() ... -> stop_scan() -> start_scan().
```

Comment:
Please add comment to describe that the `last_scanned_key` has been converted into `prefix_next` key in `stop_scan` when it is not empty.

Reply:
Yes, there is already comments:

```
                    // In `stop_scan`, we will `next(last_scanned_key)`, so we don't need to next()
                    // again here.
```